### PR TITLE
Added feature to control http body behavior

### PR DESCRIPTION
**Pull request by davidfowl**
- Added a flag to allow synchronous IO

### CODE REVIEW

**Comment by JunTaoLuo** [comment]
I assume there is more to this PR than just this?

**Comment by davidfowl** [comment]
> I was expecting a global flag on KestrelOptions, not a per request flag. Or do you plan to do both?

Both.

> I assume there is more to this PR than just this?

This is just adding the feature. I don't plan to push all at once. Feature first, then Kestrel change once it goes in.

File: src/Microsoft.AspNetCore.Http.Features/IHttpBodyControlFeature.cs [hunk 1]
```diff
@@ -0,0 +1,17 @@
+﻿using System;
```
Comment (JunTaoLuo) [inline src/Microsoft.AspNetCore.Http.Features/IHttpBodyControlFeature.cs]: copyright

File: src/Microsoft.AspNetCore.Http.Features/IHttpBodyControlFeature.cs [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+using System.Collections.Generic;
```
Comment (JunTaoLuo) [inline src/Microsoft.AspNetCore.Http.Features/IHttpBodyControlFeature.cs]: What are you using from these namespaces?

File: src/Microsoft.AspNetCore.Http.Features/IHttpBodyControlFeature.cs [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+using System.Text;
+
+namespace Microsoft.AspNetCore.Http.Features
+{
+    /// <summary>
+    /// Controls the IO behavior for the <see cref="IHttpRequestFeature.Body"/> and <see cref="IHttpResponseFeature.Body"/> 
+    /// </summary>
+    public interface IHttpBodyControlFeature
```
Comment (JunTaoLuo) [inline src/Microsoft.AspNetCore.Http.Features/IHttpBodyControlFeature.cs]: Not sure about the naming BodyControl. Maybe IHttpBodyOptionsFeature, IHttpBodyIOFeature, IHttpBodyCapabilitiesFeature?

**Comment by halter73** [comment]
@Tratcher Is HttpSysServer going to the same global flag?

It would be pretty simple to disallow sync IO using some middleware to wrap the request and response body streams. The perf hit won't be too bad if the wrapping stream is pooled and lazily initialized. It would also ensure that a server switch doesn't suddenly cause a bunch of Read/Write calls to fail.

**Comment by JunTaoLuo** [comment]
Still don't like the naming, but functionally it's fine.

**Comment by davidfowl** [comment]
@DamianEdwards do you have any naming suggestions?

File: src/Microsoft.AspNetCore.Http.Features/IHttpBodyControlFeature.cs (hunk 3 above)
Comment (rynowak) [inline src/Microsoft.AspNetCore.Http.Features/IHttpBodyControlFeature.cs]: Body Control sounds like the album name

**Comment by DamianEdwards** [comment]
`IHttpBodyIOControlFeature`
`IHttpIOControlFeature`
`IHttpBodyIOFeature`
`IHttpBodySyncIOControlFeature`
`IHttpRequestResponseBodySynchronousIOControlFeature`
`IHttpIOIOIOItsOffToWorkWeGoFeature`

**Pull request merged**